core/src/ledger_cleanup_service.rs

```diff
@@ -125,31 +134,35 @@ impl LedgerCleanupService {
         last_purge_slot: &mut u64,
         purge_interval: u64,
     ) -> Result<(), RecvTimeoutError> {
-        let mut root = new_root_receiver.recv_timeout(Duration::from_secs(1))?;
-        // Get the newest root
-        while let Ok(new_root) = new_root_receiver.try_recv() {
-            root = new_root;
+        let root = Self::receive_new_roots(new_root_receiver)?;
+        if root - *last_purge_slot <= purge_interval {
+            return Ok(());
         }
 
-        if root - *last_purge_slot > purge_interval {
-            let disk_utilization_pre = blockstore.storage_size();
+        let disk_utilization_pre = blockstore.storage_size();
+        info!(
+            "purge: new root: {}, last_purge: {}, purge_interval: {}, disk: {:?}",
+            root, last_purge_slot, purge_interval, disk_utilization_pre
+        );
+        *last_purge_slot = root;
+
+        let (num_shreds_to_clean, lowest_cleanup_slot, mut first_slot, total_shreds) =
+            Self::find_slots_to_clean(&blockstore, root, max_ledger_shreds);
+
+        if num_shreds_to_clean > 0 {
             info!(
-                "purge: new root: {} last_purge: {} purge_interval: {} disk: {:?}",
-                root, last_purge_slot, purge_interval, disk_utilization_pre
+                "cleaning slots {} to {}, num_shreds: {}",
+                first_slot, lowest_cleanup_slot, num_shreds_to_clean,
             );
-            *last_purge_slot = root;
 
-            let (num_shreds_to_clean, lowest_slot_to_clean, mut first_slot, total_shreds) =
-                Self::find_slots_to_clean(blockstore, root, max_ledger_shreds);
-
-            if num_shreds_to_clean > 0 {
-                debug!(
-                    "cleaning up to: {} shreds: {} first: {}",
-                    lowest_slot_to_clean, num_shreds_to_clean, first_slot
-                );
-                loop {
+            let purge_complete = Arc::new(AtomicBool::new(false));
+            let blockstore = blockstore.clone();
+            let purge_complete1 = purge_complete.clone();
+            let _t_purge = Builder::new()
+                .name("solana-ledger-purge".to_string())
```

Comment:
Do we need another thread, can we just receive from the channel between purge batches?

Reply:
Yeah we need the thread.  compact_storage() literally takes hours

Reply:
For each batch?

Reply:
yes

Reply:
We purge 256 slots at once. Is purging 256 slots taking hours?

Reply:
run_purge takes milliseconds, compact_storage takes hours

Reply:
huh. If that's the case, you wonder if it's really productive to have compaction enabled at all.

Reply:
Now that the huge backlog of slots on bv4 has been purged, I'm seeing ~20-30minute compact times but the ledger directory is well under 400GB (~338GB) with the default `--limit-ledger-size` (previously it was over 900GB).

Once I run this PR a little longer I'll give it a go on one of the API nodes (that currently have a  ~1.5TB ledger!).

As a follow-up I'll experiment with `--no-rocksdb-compaction` between two otherwise equally configured API nodes, to see what that does to the ledger size